rust/private/utils.bzl

```diff
@@ -379,3 +388,85 @@ def transform_deps(deps, make_rust_providers_target_independent):
         build_info = dep[BuildInfo] if BuildInfo in dep else None,
         cc_info = dep[CcInfo] if CcInfo in dep else None,
     ) for dep in deps] if make_rust_providers_target_independent else deps
+
+def should_encode_label_in_crate_name(package, third_party_dir):
+    """Determines if the crate's name should include the Bazel label, encoded.
+
+    Names of third-party crates do not encode the full label.
+
+    Args:
+        package (string): The package in question.
+        third_party_dir (string): The directory in which third-party packages are kept.
+
+    Returns:
+        True if the crate name should encode the label, False otherwise.
+    """
+
+    # TODO(hlopko): This code assumes a monorepo; make it work with external
+    # repositories as well.
+    return ("//" + package + "/").startswith(third_party_dir + "/")
+
+# This is a list of pairs, where the first element of the pair is a character
+# that is allowed in Bazel package or target names but not in crate names; and
+# the second element is an encoding of that char suitable for use in a crate
+# name.
+_encodings = (
```

Comment:
Is the encoded name supposed to be decodable, or just human readable?

Reply:
I think only hard requirement is unique for a given label, not conflicting with a different label. Decodable/human readable is nice to have, but not required.

Reply:
This encoding lets degenerate naming clash, eg. 
```
//x_slash_y:z => x_slash_y_colon_z
//x/y:z       => x_slash_y_colon_z
```

I feel like a more naive encoding (with potentially more degenerate naming clashes) would be just as good and removes the need for a macro to resolve labels.
```
//x_slash_y:z => x__slash__y_z
//x/y:z       => x_y_z
//x/y/z       => x_y_z
//x/y/z:a     => x_y_z_a
```


A bit separately, I also feel like supporting a bunch of special characters in the label names is unnecessary. We may as well go the same route as python, which requires labels to be valid python module names (eg. dashes are not importable).

If the encoding must be unambiguous, and support the additional characters, then this should use something like base64 encoding instead.

Reply:
> This encoding lets degenerate naming clash, eg.

That's not true, actually. This encoding is invertible (in the [mathematical sense](https://en.wikipedia.org/wiki/Inverse_function)). There are unit tests in utils_test.bzl to demonstrate this (though I'm aware that a test is not a proof :)). I've added the cases you mentioned, to show that those namings do not clash under this encoding.

> I feel like a more naive encoding (with potentially more degenerate naming clashes) would be just as good and removes the need for a macro to resolve labels.

For our use case, avoiding name collisions is a hard requirement. Supporting the full range of possible Bazel labels is also a hard requirement. So for us, the encoding/decoding steps are both necessary.

> If the encoding must be unambiguous, and support the additional characters, then this should use something like base64 encoding instead.

I did consider base64 encoding, but decided against it, for a few reasons:
* Rust identifiers cannot start with a digit, so we'd have to artificially modify the base64-encoded name to make sure it started with a letter.
* Rust identifiers cannot include `=` characters, so we'd have to strip out any padding bytes.
* Rust identifiers cannot include `+`, `-`, or `/` characters (which are used by the [Base64 and Base64Url alphabets](https://datatracker.ietf.org/doc/html/rfc4648#section-5)), so we'd have to do a global search/replace for those chars after encoding anyway, with some better alternative. But note that all the other characters that are legal in a crate name are already in use in the base64 alphabet, so we'd have to do some multi-character encoding. (So the base64 approach is starting to look more like the current approach.)
* Base64-encoded crate names would show up when viewing the macro-expanded version of the Rust code, but would be entirely opaque to the user. It would be better to show the user something easier to understand.

For those reasons, I decided the current approach was simpler and better from a user experience perspective as well.